### DISCUSSION TITLE
Strain gradient plasticity with `mgis.fenics`

### DISCUSSION THREAD

**Discussion by jeanmichelscherer** [thelfer/MFrontGenericInterfaceSupport#110] [Q&A]
Hi,

I am working on a scalar strain gradient plasticity extension of a simple von Mises plasticity behaviour. I would like to use that behaviour in `FEniCS` through `mgis.fenics`. The `Mfront` implementation and `FEniCS` script are attached.

The behaviour defines 3 `Gradients` (`TotalStrain`, `CumulatedPlasticStrain`, `CumulatedPlasticStrainGradient`) and 3 `Fluxes` (`Stress`, `MicroStress`, `DoubleStress`). Based on a `Mtest` test, that implementation seems to work fine.

In the `FEniCS` script, the degrees of freedom (`u=Displacement`, `lX=LagrangeMultiplier`, `pX=CumulatedPlasticStrain`) are defined through a mixed function space. Since the model does not fit into the default residual formulation, I updated the residual as follows:

```
problem.residual = ( inner(voigt2stress(sigma,dim),grad(u_)) + dot(M,grad(pX_)) + S*pX_ + (pX - p)*lX_ )*problem.dx
```

- If I pass `v`, the function defined on the mixed function space, to initialize the `MFrontNonlinearProblem` I get the following `UFL` error:

```
This integral is missing an integration domain.
Traceback (most recent call last):
  File "vonMisesStrainGradientPlasticityLagrangeMultiplier.py", line 156, in <module>
    problem.initialize()
  File "/usr/local/lib/python3.8/site-packages/mgis/fenics/nonlinear_problem.py", line 386, in initialize
    self.define_form()
  File "/usr/local/lib/python3.8/site-packages/mgis/fenics/nonlinear_problem.py", line 178, in define_form
    self.residual = (sum([
  File "/home/scherer/Softwares/ANACONDA/anaconda3/envs/fenics/lib/python3.8/site-packages/ufl/measure.py", line 441, in __rmul__
    error("This integral is missing an integration domain.")
  File "/home/scherer/Softwares/ANACONDA/anaconda3/envs/fenics/lib/python3.8/site-packages/ufl/log.py", line 172, in error
    raise self._exception_type(self._format_raw(*message))
ufl.log.UFLException: This integral is missing an integration domain.
```
- If, however, I pass `u`, only the displacement part of `v`, the issue above is not triggered.

- When I pass `u` to initialize the `MFrontNonlinearProblem`, the rest of the script runs, but the boundary conditions are not applied.

Do you know how I can pass the boundary conditions correctly?

Thanks,
Jean-Michel

[vonMisesStrainGradientPlasticity.tar.gz](https://github.com/thelfer/MFrontGenericInterfaceSupport/files/10362586/vonMisesStrainGradientPlasticity.tar.gz)

**Comment by thelfer** [comment]
@bleyerj This one is for you !

**Reply by jeanmichelscherer** [reply]
Hi @bleyerj,

The following links helped to solve the issue:
- https://thelfer.github.io/mgis/web/mgis_fenics_monolithic_transient_thermoelasticity.html
- https://fenicsproject.org/qa/11964/petsc-problem-when-solving-nonlinear-variational-problem/

I was using the wrong `split` function. The following lines were causing the issue:
```
v = Function(V, name="DOFs")
u, lX, pX = v.split()
```
They must be replaced by:
```
v = Function(V, name="DOFs")
u, lX, pX = split(v)
```
